### DISCUSSION TITLE
PorousFlow: convergence problem.

### DISCUSSION THREAD

**Discussion by sinaomrani96** [idaholab/moose#24717] [Q&A Modules: General]
Hi,
Hope you are doing well.

I am working with a 2-phase 3-component system. The input file works fine, but when I change it to compositional using PorousFlowFluidState (calculating dissolution and mass fraction in each phase) I face convergence error. I checked the Jacobian and this is the results:

No errors detected. :-)

Kernel for variable 'zi':
  (1,1) On-diagonal Jacobian is wrong (off by 62.6 %)

Kernel for variable 'tracer':
  (2,1) Off-diagonal Jacobian for variable 'zi' is slightly off (by 0.045432 %)

I was wondering if there is a problem in my input file or anything else. Thanks.

Best regards,
Sina

**Comment by sinaomrani96** [comment]
Also, I am getting 0 residual on one of the variables event though it has initial condition defined. I do not understand why this 0 residual happens. 

pwater: 0
zi:     0.000545419
tracer: 3.84182e-06

**Reply by sinaomrani96** [reply]
Hi everyone,

Sorry to ask again, but I was not able to solve it by my own. I would appreciate if you could help me. Thanks.

**Reply by GiudGiud** [reply]
@cpgr @RLiu2020 any hint on this?

**Reply by cpgr** [reply]
Sorry - I didn't see this one when I last checked.

At a guess it is a BC causing this, or might even be this https://github.com/idaholab/moose/discussions/23197#discussioncomment-4733031

I'd be happy to take a look at your example if you can share a simple version?

**Reply by sinaomrani96** [reply]
Hi Chris,
Hope you are doing well.

I checked the solution you mentioned but it did not work. About BCs, I did not change them compare to the example I had with 2-phase 3-components but without dissolution. Below is the input file link. I would appreciate it if you could look at it. Thanks.

https://github.com/sinaomrani96/PorousFlow-test/tree/main

**Reply by sinaomrani96** [reply]
Hi Chris,

I was working on the input file and now I do not get 0 residual for pressure and the jacobian analyser gives no error. However, I still cannot run it and get the following error:

Time Step 1, time = 1, dt = 1
 0 Nonlinear |R| = 5.470420e-07
    |residual|_2 of individual variables:
                     pwater: 5.47042e-07
                     zi:     0
                     tracer: 3.84182e-10
      0 Linear |R| = 5.470420e-07
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0

I outlined the conditioning number and here is the result:
      SVD: condition number            inf, 443 of 1323 singular values are (nearly) zero
      SVD: smallest singular values: 0.000000000000e+00 6.976952038536e-109 3.109126926142e-93 6.704283696648e-92 2.280963623827e-85
      SVD: largest singular values : 1.000000000000e+00 1.000000000000e+00 1.000000000000e+00 1.000000000000e+00 1.000000000000e+00

I suspect the issue is with PorousFlowBrineCO2. I specified the liquid component but I guess because I have one more component in liquid phase it faces an issue in calculating mass fractions. I would appreciate it if you could give me your insight. Thanks.

Also, below is the input file link:
https://github.com/sinaomrani96/PorousFlow-test/blob/main/test

**Reply by cpgr** [reply]
I have been testing this and I'm pretty certain that you are correct. At least, I'm sure that it isn't a bug in the code, rather a lack of functionality for this problem. If you remove the tracer the problem runs.

At the moment, there is no way to add a tracer when using one of the classes where solubility is included. Like you suggest, the main problem is that there isn't a way to add extra mass fractions. In your example, PorousFlowBrineCO2 computes the mass fraction for NaCl in water as component 2, so you can't use that for the tracer, and there isn't a way to add a fourth component (the tracer) to the mass fraction array. I don't think it would be hard, it's just that I have never needed it so didn't add it.

I think that the way to do it would be to add an extra parameter to the `PorousFlowFluidState` material that then appends extra mass fractions to the array of mass fractions, but this would require some coding.

**Reply by cpgr** [reply]
Actually, now I'm not sure that is the best idea....

**Reply by sinaomrani96** [reply]
Dear Chris,

Thank you for your time and consideration. I will work on it and see how it can be handled. Also, I had a question regarding gas viscosity when I am using PorousFlowBrineCO2 and PorousFlowFluidState. When I look at it in Paraview, it shows the value of 1 in regions that there is no gas. I checked it in the code and it is commented that it is to prevent dividing by zero. So, I was wondering that if affects the gas movement, which probably won't, and how I can remove it for visualization. Thanks.

![v](https://github.com/idaholab/moose/assets/64276459/227f7952-7358-4dde-bc68-01dcaaf60d01)

**Reply by GiudGiud** [reply]
> SVD: condition number inf, 443 of 1323 singular values are (nearly) zero

this is a big deal. The problem is severely ill-posed, the equations are not enough to determine the degree of freedoms of all the variables. 
Once you figure out the other aspects of this, you will want to check that there is no more singular values

**Reply by cpgr** [reply]
It was a while ago, but I think that there was a floating point exception when running in dbg due to the (rho * k * kr / mu) term in the advection kernel. It doesn't have any affect on the physics, because rho = 0 in this region so this term is zero. 

It does affect the visualisation though as you have observed! One option is you can use Paraview to clip the data to gas_saturation > 0 and overlay it on the full mesh.

**Reply by cpgr** [reply]
Yeah, I think it is because there aren't enough mass fractions computed to handle all of the fluid components in this case.
